_posts/2019-04-23-autoregressive-flows/experimenting-with-autoregressive-flows.Rmd

```diff
@@ -0,0 +1,645 @@
+
+With the architecture defined as above in the GAS example, this lead to the network not making much progress. Neither did leaving the data in time domain form, with 16000 features in the joint distribution. Thus, we decided to work with the FFT coefficients computed over the complete window instead, resulting in 257 joint features.^[We quickly experimented with a higher number of FFT coefficients, but the approach did not seem that promising.]
+
+```{r}
+batch_size <- 100
+
+sampling_rate <- 16000L
+data_generator <- function(df,
+                           batch_size) {
+  
+  ds <- tensor_slices_dataset(df) 
+  
+  ds <- ds %>%
+    dataset_map(function(obs) {
+      wav <-
+        decode_wav()(tf$read_file(tf$reshape(obs$fname, list())))
+      samples <- wav$audio[ ,1]
+      
+      # some wave files have fewer than 16000 samples
+      padding <- list(list(0L, sampling_rate - tf$shape(samples)[1]))
+      padded <- tf$pad(samples, padding)
+      
+      stft_out <- stft()(padded, 16000L, 1L, 512L)
+      magnitude_spectrograms <- tf$abs(stft_out) %>% tf$squeeze()
+    })
+  
+  ds <- ds  %>%
+    dataset_batch(batch_size)
+  
+}
+
+ds_train <- data_generator(df_train, batch_size)
+batch <- ds_train %>% 
+  make_iterator_one_shot() %>%
+  iterator_get_next()
+
+dim(batch) # 100 x 257
+
+```
+
+Training then proceeded as on the GAS dataset.
+
+```{r}
+# define MAF
+base_dist <-
+  tfd_multivariate_normal_diag(loc = rep(0, dim(batch)[2]))
+
+num_hidden <- 512 
+use_batchnorm <- TRUE
+use_permute <- TRUE
+num_mafs <- 10 
+num_layers <- 3 * num_mafs
+
+bijectors <- vector(mode = "list", length = num_layers)
+
+for (i in seq(1, num_layers, by = 3)) {
+  maf <- tfb_masked_autoregressive_flow(
+    shift_and_log_scale_fn = tfb_masked_autoregressive_default_template(
+      hidden_layers = list(num_hidden, num_hidden),
+      activation = tf$nn$tanh,
+      ))
+  bijectors[[i]] <- maf
+  if (use_batchnorm)
+    bijectors[[i + 1]] <- tfb_batch_normalization()
+  if (use_permute)
+    bijectors[[i + 2]] <- tfb_permute((dim(batch)[2] - 1):0)
+}
+
+if (use_permute) bijectors <- bijectors[-num_layers]
+flow <- bijectors %>%
+  discard(is.null) %>%
+  rev() %>%
+  tfb_chain()
+
+target_dist <- tfd_transformed_distribution(distribution = base_dist,
+                                            bijector = flow)
+
+optimizer <- tf$train$AdamOptimizer(1e-3)
+
+# train MAF
+n_epochs <- 100
+for (i in 1:n_epochs) {
+  agg_loglik <- 0
+  num_batches <- 0
+  iter <- make_iterator_one_shot(ds_train)
+  until_out_of_range({
+    batch <- iterator_get_next(iter)
+    loss <-
+      function()
+        - tf$reduce_mean(target_dist %>% tfd_log_prob(batch))
+    optimizer$minimize(loss)
+    
+    loglik <- tf$reduce_mean(target_dist %>% tfd_log_prob(batch))
```

Comment:
Is there a consistent style you're following to decide whether to use `%>%` or not and whether you use a newline?

Reply:
not really I guess, apart from probably a general trade-off between breaking after `%>%` but on the other hand, trying to keep code snippets as dense as possible (and probably, trying to match how much space code takes up to "how important it is" logically, i.e. trying not to have logging statements take up too much space...)
Do you have a more general suggestion (I could follow that in upcoming posts)?